api/v1alpha1/capiprovider_types.go

```diff
@@ -0,0 +1,133 @@
+/*
+Copyright SUSE 2023.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	operatorv1 "sigs.k8s.io/cluster-api-operator/api/v1alpha2"
+)
+
+const (
+	// ProviderFinalizer is the finalizer apply on the CAPI Provider resource.
+	ProviderFinalizer = "capiprovider.turtles.cattle.io"
+)
+
+// CAPIProviderSpec defines the desired state of CAPIProvider.
+// +kubebuilder:validation:XValidation:message="CAPI Provider version should be in the semver format",rule="!has(self.version) || self.version.matches(r\"\"\"^([0-9]+)\\.([0-9]+)\\.([0-9]+)(?:-([0-9A-Za-z-]+(?:\\.[0-9A-Za-z-]+)*))?(?:\\+[0-9A-Za-z-]+)?$\"\"\")"
+//
+//nolint:lll
+type CAPIProviderSpec struct {
+	// Name is the name of the provider to enable
+	// +required
+	// +kubebuilder:validation:Enum=aws;azure;gcp;docker;rke2
+	Name string `json:"name"`
+
+	// Type is the type of the provider to enable
+	// +required
+	// +kubebuilder:validation:Enum=infrastructure;core;controlPlane;bootstrap;addon
+	Type string `json:"type"`
+
+	// Credentials is the structure holding rancher credentials reference
+	// +optional
+	Credentials *ProviderCredentials `json:"credentials,omitempty"`
+
+	// Features is a collection of features to use by this provider
+	Features *Features `json:"features,omitempty"`
+
+	// Variables is a map of environment variables to add to the content of the ConfigSecret
+	// +optional
+	Variables map[string]string `json:"variables"`
+
+	// ProviderSpec is the spec
+	ProviderSpec *operatorv1.ProviderSpec `json:",inline"`
+}
+
+// Features defines a collection of features for the CAPI Provider to apply.
+type Features struct {
+	// MachinePool is an abbreviation to EXP_MACHINE_POOL: "true" feature
+	MachinePool bool `json:"machinePool,omitempty"`
+
+	// ClusterResourceSet is an abbreviation to EXP_CLUSTER_RESOURCE_SET: "true" feature
+	ClusterResourceSet bool `json:"clusterResourceSet,omitempty"`
+
+	// ClusterTopology is an abbreviation to CLUSTER_TOPOLOGY: "true" feature
+	ClusterTopology bool `json:"clusterTopology,omitempty"`
+}
+
+// ProviderCredentials defines the external credentials information for the provider.
+type ProviderCredentials struct {
```

Comment:
If **ProviderCredentials** is supplied then one of the methods would need to be used. I'm wondering if we do validation in the webhook instead of marking the **RancherCloudCredential** field required.

Reply:
Actually default kubebuilder validation is enough here to define a union. FIxed.